packages/pglite/src/pglite.ts

```diff
@@ -562,39 +562,52 @@ export class PGlite implements PGliteInterface {
   }
 
   /**
-   * Execute a postgres wire protocol message
+   * Execute a postgres wire protocol message directly without wrapping the response
    * @param message The postgres wire protocol message to execute
-   * @returns The result of the query
+   * @returns The direct message data response produced by Postgres
    */
-  async execProtocol(
+  async execProtocolRaw(
     message: Uint8Array,
     { syncToFs = true }: ExecProtocolOptions = {},
-  ): Promise<Array<[BackendMessage, Uint8Array]>> {
-    return await this.#executeMutex.runExclusive(async () => {
-      const msg_len = message.length;
-      const mod = this.mod!;
+  ) {
+    const msg_len = message.length;
+    const mod = this.mod!;
 
-      // >0 set buffer content type to wire protocol
-      // set buffer size so answer will be at size+0x2 pointer addr
-      mod._interactive_write(msg_len);
+    // >0 set buffer content type to wire protocol
+    // set buffer size so answer will be at size+0x2 pointer addr
+    mod._interactive_write(msg_len);
 
-      // copy whole buffer at addr 0x1
-      mod.HEAPU8.set(message, 1);
+    // copy whole buffer at addr 0x1
+    mod.HEAPU8.set(message, 1);
 
-      // execute the message
-      mod._interactive_one();
+    // execute the message
+    mod._interactive_one();
 
-      if (syncToFs) {
-        await this.#syncToFs();
-      }
+    if (syncToFs) {
+      await this.#syncToFs();
+    }
```

Comment:
Because of this async call between the `_interactive_one();` and `_interactive_read();` and no mutex on the  `execProtocolRaw` you could interweave calls and start a new call before the previous ended, reading the wrong result buffer value.

The #executeMutex goes back to when we read the results of the query async in an old version, but after the recent refactor and looking at the code now, I think we may be able to drop it completely (from the main execProtocol) if we move this `this.#syncToFs();` call to after `mod._interactive_read();`

syncToFs is mostly for use with the IDB VFS and flushing to disk after reading the result from the Postgres buffer, but before returning it to the user, is fine. Other VFSs (Node, and the new OPFS one) will fsync during the command as needed.

Reply:
Makes sense, updating now 👍

Quick question - this function no longer needs to be async technically, but we could keep it async if you want in case you decide you need to make this async in the future (ie. an Asyncify variant or something). Do you have a preference?

Reply:
We should keep it async as we still need to await the syncToFs call before returning to the user - just after the read from the Postgres buffer.

We could make a sync version of the method that doesn't call the syncToFs method, and leaves it to the user to manage. Could help for very efficient server use, but that's exposing some complexity we maybe shouldn't.

Reply:
Whoops - realizing now you meant to keep `this.#syncToFs();` in this method (vs. moving to `execProtocol`) which means we need to keep this async. Disregard the last question!